core/src/main/java/org/apache/accumulo/core/conf/Property.java

```diff
@@ -894,6 +893,21 @@
   TABLE_COMPACTION_STRATEGY_PREFIX("table.majc.compaction.strategy.opts.", null,
       PropertyType.PREFIX,
       "Properties in this category are used to configure the compaction strategy.", "1.6.0"),
+  // Crypto-related properties
+  @Experimental
+  TABLE_CRYPTO_PREFIX("table.crypto.opts.", null, PropertyType.PREFIX,
+      "Properties related to on-disk file encryption.", "2.1.0"),
+  @Experimental
+  @Sensitive
+  TABLE_CRYPTO_SENSITIVE_PREFIX("table.crypto.opts.sensitive.", null, PropertyType.PREFIX,
+      "Sensitive properties related to on-disk file encryption.", "2.1.0"),
+  @Experimental
+  TABLE_CRYPTO_SERVICE("table.crypto.service",
+      "org.apache.accumulo.core.spi.crypto.NoCryptoService", PropertyType.CLASSNAME,
+      "The class which executes on-disk table encryption/decryption. The default does "
+          + "nothing. This property must be a classname with an implementation of the "
+          + "org.apache.accumulo.core.spi.crypto.CryptoService interface.",
+      "2.1.0"),
```

Comment:
If there were a crypto service factory specified as an `instance.` option, then there's no need to specify a specific crypto service here, as that would be determined by the factory, given the TABLE scope and the options (either the options in the config, in the case of writing a new file, or the options read from the file in the case of reading a previously written file).

Reply:
Just so we are on the same page, you are saying we should just have one `instance.crypto.service` for the whole instance? Then if you want to specify what tables to use it, say just have a list the tables to encrypt?
`instance.crypto.tables=table1,table2,table3`
`instance.crypto.wal.enabled=true`

I think that would be easier then having something like:
`instance.crypto.service.table1=true`
`instance.crypto.service.table2=true`

Reply:
No. I'm saying we should have one `instance.crypto.service.factory`, and that should spit out crypto services based on the per-table config. This is essentially how the VolumeManager per-table feature works.

Reply:
We would use the `table.custom` the same way? The volume chooser uses the same class type for `general` properties and for `tables.custom` properties. And what about the wals? Would they have their own factory? Maybe if you give an example of your idea of a config that would help.

Reply:
I'm not a fan of the `.custom` property prefix naming convention. It implies that we have a bag of properties segregated between "official" properties in Accumulo property namespaces, and "unofficial" properties in a designated namespace for user-provided plugins. But, our reference implementations use the `.custom` property namespace in some cases but not in others. We're not consistent. I think we should just treat it as one big bag of well-named and well-scoped properties and relax the constraint that requires users to use the `.custom` namespace for their properties.

So, that said, I'd be fine leaving the `table.crypto.`, rather than require they be in `table.custom.crytpo.` or something similar. I'm not suggesting copying the property naming convention from VolumeChooser, just the pattern of letting the single factory instance decide whether and how to use these properties, based on its implementation and the provided scope.

So, imagine a `MyCryptoServiceFactoryThatSupportsPerTableConfig` being set as `instance.crypto.service.factory`. The factory has an API that is called, something like `factory.getCryptoService(scopeConfig)`. So, depending on scope, you'd provide different configs:

```java
var walCryptoService = factory.getCryptoService(systemConf.getAllPropertiesWithPrefixStripped("instance.crypto.wal."));
var tableCryptoServiceForTableA = factory.getCryptoService(tableAConf.getAllPropertiesWithPrefixStripped("table.crypto."));
```

Now, it's up to the implementation of the factory to decide what CryptoService to return, based on the scope and scope config. (My example only shows passing in config properties, but you could pass in the scope as a type as well, or some other object performing a similar function of communicating context to the factory.)

Reply:
OK. Yeah I like `table.crypto.` better as well. 

If we are not specifying the crypto service in the table config, do we use the same crypto service for all tables? If not, what in the table config will determine which crypto service to use. Based on what you are saying, I am imagining a config like this:
<pre>
instance.crypto.factory=MyCryptoServiceFactoryThatSupportsPerTableConfig
instance.wal.enabled=true
instance.crypto.wal.opts.key.uri=/home/mike/workspace/uno/CryptoTest-testkeyfile
secureTable:
table.crypto.enabled=true
table.crypto.type=strong
table.crypto.opts.key=/something/more/secure/than/a/file
lessSecureTable:
table.crypto.enabled=true
table.crypto.type=weak
table.crypto.opts.key.uri=/home/mike/workspace/uno/CryptoTest-testkeyfile2
</pre>

With the impl of MyCryptoServiceFactoryThatSupportsPerTableConfig looking like this:
<pre>
if scope == wal && enabled, return AESCryptoService
if scope == table && enabled && type=strong, return StrongerCryptoService
if scope == table && enabled && type=weak, return WeakerCryptoService
</pre>

Reply:
> If we are not specifying the crypto service in the table config, do we use the same crypto service for all tables? If not, what in the table config will determine which crypto service to use.

I think that would be implementation-dependent.

> Based on what you are saying, I am imagining a config like this:
> 
> instance.crypto.factory=MyCryptoServiceFactoryThatSupportsPerTableConfig
> instance.wal.enabled=true
> instance.crypto.wal.opts.key.uri=/home/mike/workspace/uno/CryptoTest-testkeyfile
> secureTable:
> table.crypto.enabled=true
> table.crypto.type=strong
> table.crypto.opts.key=/something/more/secure/than/a/file
> lessSecureTable:
> table.crypto.enabled=true
> table.crypto.type=weak
> table.crypto.opts.key.uri=/home/mike/workspace/uno/CryptoTest-testkeyfile2
> With the impl of MyCryptoServiceFactoryThatSupportsPerTableConfig looking like this:
> 
> if scope == wal && enabled, return AESCryptoService
> if scope == table && enabled && type=strong, return StrongerCryptoService
> if scope == table && enabled && type=weak, return WeakerCryptoService

Something like that. I would probably just put everything under `table.crypto.opts.`, though. So, something closer to:

```ini
instance.crypto.service.factory=PerTableAESCryptoFactory
instance.crypto.wal.opts.algorithm=AES256+CBC
instance.crypto.wal.opts.key.uri=/path/to/keyfile
; no crypto table (still provide the key, in case it needs to read old data):
table.crypto.opts.algorithm=NONE
table.crypto.opts.key.uri=/path/to/keyfile
; table A:
table.crypto.opts.algorithm=AES128+ECB
table.crypto.opts.key.uri=/path/to/keyfile
; table B:
table.crypto.opts.algorithm=AES512+ECB
table.crypto.opts.key.uri=/path/to/keyfile
```

inside PerTableAESCryptoFactory
```java
  if (algorithm == NONE) return new CryptoService(new NoEncrypter(), new AESDecrypter(key));
  else return new CryptoService(new AESEncrypter(algorithm, key), new AESDecrypter(key));
```

However, exactly how/where the algorithm/key/other opts get passed depends on what the interfaces look like exactly, and at what point are they called.

Reply:
> So it looks like these are the only crypto properties we need to bake into Accumulo's Property.java:

Looks mostly right to me. But, I'm not sure about the "sensitive" stuff. I think maybe we could just assume all crypto parameters are "sensitive". This really only has to do with returning the config via the public API for printing in the shell. We could require admin permission to retrieve/view any of those without creating two maps for all the opts.

> instance.crypto.factory (don't think we need service in the name)

Agreed on the parenthetical.

> Do you think we need to keep the two we have now `instance.crypto.opts.` and `instance.crypto.opts.sensitive.`? The only thing I can think of is if a user needs generic properties for the factory itself.

I'm not sure. If you use them, then the interface would probably have to require an `init(Map<String,String> opts, Map<String,String> sensitiveOpts)` method, in addition to a no-arg constructor to instantiate the class. (or just a single map parameter, if we drop the separate set of sensitive props).